### PR TITLE
notification integration with Socket 

### PR DESCRIPTION
**Pull request by serafy0**
### What this PR does (required):
- integrates notifications with socket
- whenever you post a new `request` or `notification` the sockets should emit the new notification to user by his id

### Screenshots / Videos (required):

https://user-images.githubusercontent.com/29197126/133566063-acfdbcc3-fa13-4ba7-b473-f2e37250b6f5.mov

https://user-images.githubusercontent.com/29197126/133566072-766ec9f4-8ce1-4704-86ed-d45faa0c4fc4.mov

### Any information needed to test this feature (required):
- post a new `/request` or `/notification` routes 

### Any issues with the current functionality (optional):
- I'm not sure what's the correct way to push to an array of interface objects a new message using typescript and react hook
- i tried this code but it returns the error message bellow but it only works with `@ts-ignore`
```js 
setNotifications((oldNotificationsList) => [newNotification, ...oldNotificationsList]);
```

- `TS2345: Argument of type '(oldNotificationsList: [Notification] | undefined) => [Notification, ...any[]]' is not assignable to parameter of type 'SetStateAction<[Notification] | undefined>'.   Type '(oldNotificationsList: [Notification] | undefined) => [Notification, ...any[]]' is not assignable to type '(prevState: [Notification] | undefined) => [Notification] | undefined'.     Type '[Notification, ...any[]]' is not assignable to type '[Notification]'.       Target allows only 1 element(s) but source may have more.`

### CODE REVIEW

**Comment by bonnieli** [comment]
> i tried this code but it returns the error message bellow but it only works with @ts-ignore
setNotifications((oldNotificationsList) => [newNotification, ...oldNotificationsList]);
TS2345: Argument of type '(oldNotificationsList: [Notification] | undefined) => [Notification, ...any[]]' is not assignable to parameter of type 'SetStateAction<[Notification] | undefined>'.   Type '(oldNotificationsList: [Notification] | undefined) => [Notification, ...any[]]' is not assignable to type '(prevState: [Notification] | undefined) => [Notification] | undefined'.     Type '[Notification, ...any[]]' is not assignable to type '[Notification]'.       Target allows only 1 element(s) but source may have more.


To address this, it looks like the new list just doesn't follow the proper type than an array of `Notification`. When you debug this, what do you expect and what are you actually receiving?

**Comment by serafy0** [comment]
I've tried to debug this and I think the type error was because typescript defines `[Notifiaction]` as a tuple where I changed it to `Notification[]` at it worked

File: server/bin/www [hunk 1]
```diff
@@ -34,25 +34,25 @@ const authSocket = async (socket, next) => {
   console.log(`token ${JSON.stringify(token)}`);
   const { username } = auth;
   if (!token) {
-    return res.status(401).send('No token, authorization denied');
-  }
+    next(new Error('No token, authorization denied'));
```
Comment (ezeilo-su) [inline server/bin/www:37]: Nice use of error middleware

File: server/controllers/request.js [hunk 2]
```diff
@@ -70,6 +70,8 @@ exports.postRequest = asyncHandler(async (req, res, next) => {
     },
   });
 
+  req.io.to(sitterId).emit('new-notification', newNotification);
```
Comment (ezeilo-su) [inline server/controllers/request.js:73]: :+1:

File: server/bin/www (hunk 1 above)
Comment (ezeilo-su) [inline server/bin/www:37]: Use proper indentation

**Review by ezeilo-su — COMMENTED** [review]
Well done for implementing this one.
Keep your code clean by commenting out lines of code used for debugging, removing unused codes, and using linters properly to format code.

**Review by moffatethan — COMMENTED** [review]
Good work. Left you a comment to address 😄

**Review by moffatethan — APPROVED** [review]

**Pull request merged**